### PR TITLE
run: deprecate '--local' flag in favor of 'local' subcommand

### PR DESCRIPTION
**Pull request by estroz**
**Description of the change:**
* cmd/operator-sdk/run: deprecate 'run --loca' and related flags, and move 'local' and 'packagemanifests' subcommand code into their own packages
* *: s/--local/local/g in docs

**Motivation for the change:** follows the pattern of [`run packagemanifests`](#3016) and [`run bundle](https://github.com/operator-framework/enhancements/pull/19).

/kind feature

### CODE REVIEW

File: website/content/en/docs/helm/quickstart.md [hunk 1]
```diff
@@ -214,7 +214,7 @@ INFO[0000] operator-sdk Version: v0.1.1+git
 Run the operator locally with a provided Kubernetes config file:
 
 ```sh
-$ operator-sdk run --local --kubeconfig=<path_to_config>
+$ operator-sdk run local --kubeconfig=<path_to_config>
```
Comment (camilamacedo86) [inline website/content/en/docs/helm/quickstart.md:217]: Since the kubeconfig was depreciated here should we keep it in the doc?

File: changelog/fragments/run-local.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+entries:
+  - description: >
+      Add 'run local' subcommand, which has the same functionality of the
+      deprecated 'run --local' mode.
```
Comment (camilamacedo86) [inline changelog/fragments/run-local.yaml:4]: It is doing more than that. 
We need to put here the flags deprecated and the new subcommand added as well.

File: cmd/operator-sdk/run/cmd.go [hunk 3]
```diff
@@ -60,13 +62,10 @@ func NewCmd() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "run",
 		Short: "Run an Operator in a variety of environments",
-		Long: `This command will run or deploy your Operator in two different modes: locally
-and using OLM. These modes are controlled by setting --local and --olm run mode
-flags. Each run mode has a separate set of flags that configure 'run' for that
-mode. Run 'operator-sdk run --help' for more information on these flags.
-
-Read more about the --olm run mode and configuration options here:
-https://sdk.operatorframework.io/docs/olm-integration/cli-overview
```
Comment (camilamacedo86) [inline cmd/operator-sdk/run/cmd.go:69]: Should we really remove the link for the doc which give further information? How users will know what means, I mean when an why the `packagemanifests` subcommand would be used? Or what means `using OLM`?

File: cmd/operator-sdk/run/cmd.go [hunk 4]
```diff
@@ -60,13 +62,10 @@ func NewCmd() *cobra.Command {
...
+		Long: `This command has subcommands that will run or deploy your Operator in two
+different modes: locally and using OLM. These modes are controlled by using 'local'
+or 'packagemanifests' subcommands. Run 'operator-sdk run --help' for more
```
Comment (camilamacedo86) [inline cmd/operator-sdk/run/cmd.go:67]: Will we have the run bundle as well? 
It shows a little confuse? Was not the olm command able to work with bundles?

File: cmd/operator-sdk/run/cmd.go [hunk 5]
```diff
@@ -114,25 +113,28 @@ https://sdk.operatorframework.io/docs/olm-integration/cli-overview
...
 	}
-	// Avoid sorting flags so we can group them according to run type.
-	cmd.Flags().SortFlags = false
 
 	// Shared flags.
 	cmd.Flags().StringVar(&c.kubeconfig, "kubeconfig", "",
 		"The file path to kubernetes configuration file. Defaults to location "+
 			"specified by $KUBECONFIG, or to default file rules if not set")
+	err := cmd.Flags().MarkDeprecated("kubeconfig",
+		"use --kubeconfig with 'local' or 'packagemanifests' subcommands instead")
```
Comment (camilamacedo86) [inline cmd/operator-sdk/run/cmd.go:134]: Would not make more sense has a specific command to OLM features instead of re-use the run command?
Add a new subcommand as `packagemanifests` shows not be part of the scope `deprecate the --local flag in favor of a new subcommand local`

File: cmd/operator-sdk/run/cmd.go [hunk 6]
```diff
@@ -149,38 +151,48 @@ https://sdk.operatorframework.io/docs/olm-integration/cli-overview
 	cmd.Flags().BoolVar(&c.olm, "olm", false,
 		"The operator to be run will be managed by OLM in a cluster. "+
 			"Cannot be set with another run-type flag")
-	err := cmd.Flags().MarkDeprecated("olm", "use 'run packagemanifests' instead")
+	err = cmd.Flags().MarkDeprecated("olm", "use 'run packagemanifests' instead")
```
Comment (camilamacedo86) [inline cmd/operator-sdk/run/cmd.go:154]: `deprecated olm in favor of a new subcommand` shows not part of the scope `deprecate '--local' flag in favor of 'local' subcommand. Could not it be done in another PR? If we need to revert the specific change for olm, for example,  would not be an issue?

File: cmd/operator-sdk/run/local/cmd.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	c.AddToFlags(cmd.Flags())
+	switch projutil.GetOperatorType() {
+	case projutil.OperatorTypeAnsible:
+		c.AnsibleOperatorFlags = aoflags.AddTo(cmd.Flags(), "(ansible operator)")
+	case projutil.OperatorTypeHelm:
+		c.HelmOperatorFlags = hoflags.AddTo(cmd.Flags(), "(helm operator)")
+	}
+
```
Comment (camilamacedo86) [inline cmd/operator-sdk/run/local/cmd.go:69]: Is not it missing the go type?

File: cmd/operator-sdk/run/local/local.go [hunk 8]
```diff
@@ -40,41 +40,42 @@ import (
...
-	kubeconfig           string
-	watchNamespace       string
-	operatorFlags        string
-	ldFlags              string
-	enableDelve          bool
-	ansibleOperatorFlags *aoflags.AnsibleOperatorFlags
-	helmOperatorFlags    *hoflags.HelmOperatorFlags
+type RunLocalCmd struct {
+	Kubeconfig           string
+	WatchNamespace       string
```
Comment (camilamacedo86) [inline cmd/operator-sdk/run/local/local.go:45]:
```suggestion
        // Valid just for the legacy layout
        // todo: remove when the legacy layout is no longer supported 
	WatchNamespace       string
```

File: website/content/en/docs/cli/operator-sdk_run_local.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+### Options
+
+```
+      --enable-delve             Start the operator using the delve debugger
+      --go-ldflags string        Set Go linker options
+  -h, --help                     help for local
+      --kubeconfig string        The file path to kubernetes configuration file. Defaults to location specified by $KUBECONFIG, or to default file rules if not set
+      --operator-flags string    The flags that the operator needs. Example: "--flag1 value1 --flag2=value2"
+      --watch-namespace string   The namespace where the operator watches for changes. Set "" for AllNamespaces, set "ns1,ns2" for MultiNamespace
```
Comment (camilamacedo86) [inline website/content/en/docs/cli/operator-sdk_run_local.md:25]: Should we not make clear that the watch-namespace is not valid for the new layout?

File: cmd/operator-sdk/run/cmd.go (hunk 4 above)
Comment (estroz) [reply cmd/operator-sdk/run/cmd.go:67]: It could theoretically work with bundles but is not recommended. Once `run bundle` is added we can update this description.

File: cmd/operator-sdk/run/cmd.go (hunk 3 above)
Comment (estroz) [reply cmd/operator-sdk/run/cmd.go:69]: This link is in `run packagemanifests --help` now.

File: cmd/operator-sdk/run/cmd.go (hunk 6 above)
Comment (estroz) [reply cmd/operator-sdk/run/cmd.go:154]: This wasn't added, I only changed `:=` to `=`.

File: cmd/operator-sdk/run/local/cmd.go (hunk 7 above)
Comment (estroz) [reply cmd/operator-sdk/run/local/cmd.go:69]: There are no special flags for `run local` on Go operator types.

File: cmd/operator-sdk/run/local/local.go (hunk 8 above)
Comment (estroz) [reply cmd/operator-sdk/run/local/local.go:45]: Seeing as projects created with `init` can be namespaced (with extra work), I don't see a reason to ever remove this.

File: website/content/en/docs/cli/operator-sdk_run_local.md (hunk 9 above)
Comment (estroz) [reply website/content/en/docs/cli/operator-sdk_run_local.md:25]: See https://github.com/operator-framework/operator-sdk/pull/3067/files#r427374500. This description should be changed to be specific to `run local` though, since the current one is geared towards OLM.

File: cmd/operator-sdk/run/local/local.go (hunk 8 above)
Comment (estroz) [reply cmd/operator-sdk/run/local/local.go:45]: Furthermore do we want to disable setting `--watch-namespaces`, as we currently do, for new project layouts, considering that they can be namespaced @joelanford @hasbro17? Out of scope for this PR, still worth discussing.

File: cmd/operator-sdk/run/cmd.go (hunk 4 above)
Comment (camilamacedo86) [reply cmd/operator-sdk/run/cmd.go:67]: cacher the idea .. tks for the clarification.

File: cmd/operator-sdk/run/local/local.go (hunk 8 above)
Comment (camilamacedo86) [reply cmd/operator-sdk/run/local/local.go:45]: Hi @estroz,  It is disabled for kb layout already. I understand that so far we do not think in port this for kb layouts. 
However, may we can plan (create a task) to check if has any reason that would make it valuable be ported.

File: cmd/operator-sdk/run/cmd.go (hunk 4 above)
Comment (camilamacedo86) [reply cmd/operator-sdk/run/cmd.go:67]: I am ok with here in this PR. It is may out of scope. However, since I just checked these here then, I'd like to share the thought.  

We are working on to have: 

- operator-sdk run packagemanifests -> run olm project integration with pkg manifest config 
- operator-sdk run  bundle -> run olm project  integration with pkg manifest config 
- operator-sdk run local -> run the project locally

However, may not make more sense something as : 

- operator-sdk run olm [OPTIONS: --config='pkg' ] -> run olm project integration  where the default config is by using bundle 
- operator-sdk run local -> run the project locally

Comment (estroz) [reply cmd/operator-sdk/run/cmd.go:67]: `run bundle` and `run packagemanifests` will have different flag sets and underlying concept (bundle vs package manifests format). I think it makes sense to differentiate them by subcommand rather than run mode, and makes deprecation of package manifests easier down the line (although we have no concrete plan to deprecate).

File: cmd/operator-sdk/run/local/local.go (hunk 8 above)
Comment (estroz) [reply cmd/operator-sdk/run/local/local.go:45]: It is currently disabled, and I'm wondering if we want to continue to do that. If a new project layout is made to be namespaced (manager consumes `--namespaces` or `WATCH_NAMESPACE`), shouldn't they be able to use `--watch-namespaces`?

Comment (camilamacedo86) [reply cmd/operator-sdk/run/local/local.go:45]: The WATCH_NAMESPACE feature requires changes in the scaffold which are not made in the kb layout so was decided that until we have sure that we would like to migrate this feature and it that it would be really valuable the flag would be hidden and not "officially" deprecated. See the comments in the code; https://github.com/operator-framework/operator-sdk/pull/3067/files#diff-3b91b8e7a6ffd108dcc88e06b3c11a01R41-R43

PS: Since it is related to the operator scope setup options as well. I added in the operator scope doc the steps which would be required to have the code changes that also could allow the flags work. See: https://github.com/operator-framework/operator-sdk/pull/2982/files#diff-f0fb41913d0dbfca6bcc82359ab92aacR113-R185

Comment (estroz) [reply cmd/operator-sdk/run/local/local.go:45]: We can address this in a follow-up.

File: cmd/operator-sdk/run/cmd.go (hunk 4 above)
Comment (camilamacedo86) [reply cmd/operator-sdk/run/cmd.go:67]: Fair enough

File: cmd/operator-sdk/run/local/local.go [hunk 10]
```diff
@@ -40,41 +40,42 @@ import (
...
 	// The main.go and manager.yaml scaffolds in the new layout do not support the WATCH_NAMESPACE
 	// env var to configure the namespace that the operator watches. The default is all namespaces.
 	// So this flag is unsupported for the new layout.
 	if !kbutil.HasProjectFile() {
-		fs.StringVar(&c.watchNamespace, "watch-namespace", "",
-			prefix+"The namespace where the operator watches for changes. Set \"\" for AllNamespaces, "+
-				"set \"ns1,ns2\" for MultiNamespace")
+		fs.StringVar(&c.WatchNamespace, "watch-namespace", "",
+			"The namespaces where the operator watches for changes. "+
+				"Set this flag if your operator is not cluster-scoped")
```
Comment (camilamacedo86) [inline cmd/operator-sdk/run/local/local.go:60]: Should not change the description here of the flag and it is NOT accurate.

Comment (estroz) [reply cmd/operator-sdk/run/local/local.go:60]: Using this terminology is weird if not using OLM, which this command does not. I will change this back and make a follow up PR where we can discuss this.
